src/blas/impl/KokkosBlas2_serial_gemv_internal.hpp

```diff
@@ -0,0 +1,155 @@
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+#ifndef __KOKKOSBLAS_GEMV_SERIAL_INTERNAL_HPP__
+#define __KOKKOSBLAS_GEMV_SERIAL_INTERNAL_HPP__
+
+/// \author Kyungjoo Kim (kyukim@sandia.gov)
+
+#include "KokkosBlas_util.hpp"
+#include "KokkosBlas1_set_impl.hpp"
+#include "KokkosBlas1_serial_scal_impl.hpp"
+#include "KokkosBlas2_serial_gemv_inner_multiple_dot.hpp"
+
+namespace KokkosBlas {
+namespace Impl {
+///
+/// Serial Internal Impl
+/// ====================
+
+template <typename ArgAlgo>
+struct SerialGemvInternal {
+  template <typename OpA, typename ScalarType, typename ValueAType,
+            typename ValueXType, typename ValueYType>
+  KOKKOS_INLINE_FUNCTION static int invoke(
+      OpA op, const int m, const int n, const ScalarType alpha,
+      const ValueAType *KOKKOS_RESTRICT A, const int as0, const int as1,
+      const ValueXType *KOKKOS_RESTRICT x, const int xs0, const ScalarType beta,
+      /**/ ValueYType *KOKKOS_RESTRICT y, const int ys0);
+
+  // default OpA = OpID
+  template <typename ScalarType, typename ValueAType, typename ValueXType,
+            typename ValueYType>
+  KOKKOS_INLINE_FUNCTION static int invoke(
+      const int m, const int n, const ScalarType alpha,
+      const ValueAType *KOKKOS_RESTRICT A, const int as0, const int as1,
+      const ValueXType *KOKKOS_RESTRICT x, const int xs0, const ScalarType beta,
+      /**/ ValueYType *KOKKOS_RESTRICT y, const int ys0) {
+    return invoke(OpID(), m, n, alpha, A, as0, as1, x, xs0, beta, y, ys0);
+  }
+};
+
+template <>
+template <typename OpA, typename ScalarType, typename ValueAType,
+          typename ValueXType, typename ValueYType>
+KOKKOS_INLINE_FUNCTION int SerialGemvInternal<Algo::Gemv::Unblocked>::invoke(
+    OpA op, const int m, const int n, const ScalarType alpha,
+    const ValueAType *KOKKOS_RESTRICT A, const int as0, const int as1,
+    const ValueXType *KOKKOS_RESTRICT x, const int xs0, const ScalarType beta,
+    /**/ ValueYType *KOKKOS_RESTRICT y, const int ys0) {
+  const ScalarType one(1.0), zero(0.0);
```

Comment:
This could even be constexpr

Reply:
Sure, made it `constexpr` on https://github.com/kokkos/kokkos-kernels/pull/1433/commits/54e9f48ed28e8937bc87c62502cee1b0a5db2490.

Reply:
Ok... apparently it won't work with complex (see [the compiler errors](https://github.com/kokkos/kokkos-kernels/runs/7585076099?check_suite_focus=true)), so let me cancel that commit.